---
title: "Visualization"
author: 
  - name: Kevin Stachelek
    affiliation:
    - University of Southern California   
    email: kevin.stachelek@gmail.com
  - name: Bhavana Bhat
    affiliation:
    - University of Southern California   
    email: bbhat@usc.edu
output: 
  BiocStyle::html_document:
    self_contained: yes
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: show
date: "`r doc_date()`"
package: "`r pkg_ver('chevreul')`"
vignette: >
  %\VignetteIndexEntry{Visualization}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    crop = NULL, ## Related to https://stat.ethz.ch/pipermail/bioc-devel/2020-April/016656.html
    dpi = 900,
    out.width = "100%",
    message = FALSE,
    warning = FALSE,
    fig.width = 10,
    fig.height = 6
)

```

This article demonstrates the data visualization tools in Chevreul. We'll 
introduce included functions, their usage, and resulting plots 

First step is to load Chevreul package and all other packages required 

```{r packages}
library(chevreul)
library(scater)
library(scran)
library(tidyverse)
library(ggraph)
library(patchwork)

# library(chevreuldata)
# chevreul_sce <- human_gene_transcript_sce()

library(ExperimentHub)
eh <- ExperimentHub::ExperimentHub()
human_gene_transcript_sce <- eh[["EH9452"]]
```


The different plotting functions within Chevreul allows for visualization of 
data, these plots can be customized for interactive or non-interactive display.

# Plot expression 

Expression of a feature (genes or transcripts) can be plotted on a given 
embedding resulting in an interactive feature plot. 

When plotting only one feature, output is identical to 
`SingleCellExperiment::FeaturePlot` 
 
```{r}
plot_feature(chevreul_sce,
    embedding = "UMAP",
    features = "NRL", return_plotly = FALSE
)
```


An interactive output plot can be generated by specifying `return_plotly = TRUE`
which uses `ggplotly` allowing identification of individual cells for further 
investigation.

## Plot read count or other QC measurements

The `plot_readcount` function displays a histogram of cell read counts colored 
according to a categorical variable using the argument `color.by`. Here we can 
see that read counts for this dataset are not distinctly different depending on
the method of organoid preparation used (Kuwahara or Zhong)

```{r}
plot_readcount(chevreul_sce,
    group_by = "nCount_RNA",
    color.by = "Prep.Method"
)
```

## Plot metadata variable

Make an interactive scatter plot of a metadata variable, where each point in
the plot represents a cell whose position on the plot is given by the cell 
embedding determined by the dimensional reduction technique by default, "UMAP".
The group argument specifies the colData variable by which to group the cells 
by, by default, "batch".

```{r}
plot_var(chevreul_sce,
    group = "gene_snn_res.0.2",
    embedding = "UMAP"
)
```

This function utilizes a SingleCellExperiment function, `DimPlot()`, as sub 
function which produces the dimensional reduction plot. The interactive 
parameter, `return_plotly`, in plot_var when set to TRUE will convert the 
plot into an interactive plot using ggplotly function from R's plotly package

## Plot cluster marker genes  

Marker genes of louvain clusters or additional experimental metadata can be 
plotted using `plot_markers`. This allows visualization of n marker features 
grouped by the metadata of interest. Marker genes are identified using wilcoxon 
rank-sum test as implemented in `presto`. In the resulting dot plot the size 
of the dot corresponds to the percentage of cells expressing the feature in 
each cluster and the color represents the average expression level of the 
feature. 


```{r}
plot_markers(chevreul_sce,
    group_by = "gene_snn_res.0.2",
    marker_method = "wilcox"
)
```


## Violin plot

To visualize the distribution of expression level of a feature in different 
groups of cells Chevreul draws a violin plot. This function uses 
SingleCellExperiment's `VInPlot()` function as a sub-function to create 
the violin plot, where the metadata variable, provided to the function 
through the variable `plot_var`, is used to group the cells and based on the
level of feature expression a violin plot is produced. 

```{r}
plot_violin(chevreul_sce, plot_var = "Kit_ID", features = "NRL")
```


## Plotting all transcripts 

Expression level of all the transcripts present in a given SingleCellExperiment
object for a gene of interest. specified by the argument 'features'. By default 
the data is superimposed on UMAP embedding. 

```{r}
plot_all_transcripts(chevreul_sce, features = "NRL")
```


## Plotting transcript composition

`plot_transcript_composition()` plots the proportion of reads of a given gene
map to each transcript. The gene of interest is specified by the argument 
'gene_symbol'. 
 
```{r, results=FALSE, echo=FALSE}
data("grch38")
data("grch38_tx2gene")
plot_transcript_composition(chevreul_sce, "NRL",
    group.by = "gene_snn_res.0.2", standardize = TRUE
)
```


## Plot Complex heatmap from SingleCellExperiment object 

We can plot an annotated complex heat map from the SingleCellExperiment object 
with a dendogram added to the left side and the top.


```{r}
top_10_features <- getTopHVGs(chevreul_sce)[1:10]

make_complex_heatmap(chevreul_sce,
    features = top_10_features,
    group.by = "gene_snn_res.0.2"
)

make_complex_heatmap(chevreul_sce,
    features = top_10_features,
    group.by = "gene_snn_res.0.2", col_arrangement = "average"
)

make_complex_heatmap(chevreul_sce,
    features = top_10_features,
    group.by = c("Prep.Method", "gene_snn_res.0.2"),
    col_arrangement = "gene_snn_res.0.2"
)
```

## Session information

```{r reproduce3, echo=FALSE}
## Session info
library("sessioninfo")
options(width = 120)
session_info()
```
